### PR TITLE
Add trailing "/"s to keep Netlify proxying happy

### PR DESCRIPTION
**Pull request by hwillson**
Without trailing "/"s, a redirect happens when accessing spec URLs, which leads to our Netlify proxying breaking. These changes make sure https://specs.apollo.dev is used for all specs URLs.

### CODE REVIEW

**Review by abernix — APPROVED** [review]

**Comment by glasser** [comment]
How hard will it be to keep the non-slashed versions working without a redirect to the other domain?

The actual core spec itself defines the "feature URL" of a spec (the thing you put in `@core(feature: "https://specs.apollo.dev/join/v0.1")` as not ending in slashes. Or actually, hmm, I think the current core spec is a bit ambiguous about this: it's very clear that the non-version "identity" part of it (`https://specs.apollo.dev/join`) doesn't have a trailing slash, but isn't clear about whether the full URL may contain a trailing slash. (I believe both of Ashi's implementations of the core spec logic in Rust and TS enforce no trailing slash, but we're not currently using those implementation's anywhere.)

I think we should make sure that the non-slashed versions don't redirect to the other domain if feasible, though I also think perhaps we should change the core spec to make a trailing slash after the version part optional rather than banned. (And change our current non-spec-complaint [hardcoded parser in federation](https://github.com/apollographql/federation/blob/96acfefe1583ca9180e2ad9b4ca8452ef3adafbf/query-planner-js/src/composedSchema/buildComposedSchema.ts#L53-L54) to allow trailing slashes.)

Specifically I would change the part in https://apollo-specs.github.io/core/v0.1/#sec-Ignore-meaningless-URL-components which talks about `#fragments` and `?query-strings` being meaningless in feature URLs to also say that trailing slashes are meaningless.

**Comment by ndintenfass** [comment]
The thing we punted on that might be relevant here is how one would define a machine-readable URL -- it would be nice if one could reason about that relative to the feature URL and avoid having to declare two different URLs.

**Comment by glasser** [comment]
I don't think that's particularly relevant to the slash issue.

My instinct would be that the way to convert a "human URL" to a "machine-readable URL" is by putting a `<meta>` tag into the human web page HTML that links to the machine-readable version, rather than by putting more constraints on the URL syntax. (This is a pretty common way to achieve the goal of "let you use a pretty human-resolvable URL in an automated context", eg it's how [Go implements going from a package URL to the version control system behind it](https://golang.org/cmd/go/#hdr-Remote_import_paths))

**Comment by ndintenfass** [comment]
I don't love relying on a particular HTML construct (and force any code to do HTML parsing and an extra network hop), but also fine if we punt that down the road and not worry about it for the sake of this ticket.

**Comment by glasser** [comment]
Yes, I mean there's a really short term issue here in that we currently only support slashless URLs in the directives (in our code and the spec) but we aren't serving those URLs without an unattractive redirect to a different domain that we don't want to show.  This is pretty separate from questions about how a to-be-designed machine-readable format would be discovered.

**Comment by abernix** [comment]
I would propose we need a slight change to the spec in this regard, @glasser.

> How hard will it be to keep the non-slashed versions working without a redirect to the other domain?

I believe the answer might be _hard_.  Netlify normalizes — at their edge routers — to include the `/`-trailing version, claiming that it's a technique that increases the cache-hit rate.  I'm personally not the hugest fan of that (since I see `/this` and `/this/` as representing different things; it changes the relativity-resolution of URLs on one page or the other), but I can see the value they're claiming and... well, it is what it is, I guess.  (I could dig into it more, or contact their support, I suppose?)

Here's a request to Apollo Server docs, as an example:

```
$ curl -v https://apollographql.com/docs/apollo-server |& grep -i location:
< location: https://www.apollographql.com/docs/apollo-server
```

I thought for a second that this might be _because_ of the way we have defined redirect rules, but going further, it appears that GitHub Pages seems to behave the same way:

```
$ curl -v https://apollo-specs.github.io/join/v0.1 |& grep -i location:
< location: https://apollo-specs.github.io/join/v0.1/
```

... forcing the trailing `/`.

I think we should allow the trailing slash to be present?

**Comment by abernix** [comment]
With that said, while I agree with the idea that this could be represented with metadata in the HTML itself, I think I might see the relevancy Nathan is drawing to the machine-readable aspect of this and considerations that we might want to make _today_ for that.  For example, I could imagine machine-readable details working without metadata being declared in HTML, in a way similar to how `/favicon.ico` works today: if it's _not defined_ in the HTML, it still works by attempting a read from a `/known-location` on top of the base-path (e.g. `..../v1.10/known-location`).

I might claim that we require the trailing slash to denote that this is a container/root (e.g., a directory/bucket) that might have other components within it?  Regardless, requiring the slash and relaxing the rule in the future seems better to me than proceeding without it now?  (Frankly, I could also see someone merely using `https://v2.my-specs.com/`, but 🤷 
 — I think that might break something / not be possible right now?).

**Comment by glasser** [comment]
@abernix are you saying it's hard to make Netlify support "unslashed urls without a redirect" or "unslashed urls without a redirect to the GHP domain behind it"? The latter sounds very surprising to me!

For the spec issue, given that the spec expects you to parse URLs and ignore parts anyway, I'd like to just change it to make a trailing slash optional when parsing in your server. This reduces the odds that people will have to fight with their hosting environments in ways like this. (Note that I would *not* say that for the purposes of letting a human read the spec you have to let both versions resolve -- just the "identity and version extraction" should work either way.)

**Comment by abernix** [comment]
> are you saying it's hard to make Netlify support "unslashed urls without a redirect" or "unslashed urls without a redirect to the GHP domain behind it"? The latter sounds very surprising to me!

Happily take another set of eyes, but the latter is proving to be problematic. (Tried to show this above, but...)

```
$ curl -v https://specs.apollo.dev/core/v0.1 |& grep -iE '^< (location|server:)'
< server: Netlify
< location: https://apollo-specs.github.io/core/v0.1/
```

```
$ curl -v https://specs.apollo.dev/core/v0.1/ |& grep -iE '^< (location|server:)'
< server: Netlify
```

I don't quite understand why, but none of my attempts (https://github.com/apollographql/website-router/pull/9) seem to be paying off.  Unfortunately, I'm not able to go deeper into this because of time.  Does my answer dictate a difference in our path forward to you, @glasser?

**Comment by glasser** [comment]
@hwillson curious if you have any ideas here?  or do you think we could get Netlify support perhaps?

I am going to change the spec to make the slash-less URL ok, but I really think we should be able to get the domain working in a way where it totally hides the backing domain (which we might be changing)...

**Comment by abernix** [comment]
I've contacted Netlify support.

Since we do this successfully on all of our documentation repositories/deployments that aren't involving GitHub pages, I suspect this could have something to do with proxying to an external-to-Netlify domain.  I'm almost certain we're experiencing this because GitHub is doing the trailing-slash normalization on their own un-slashed pages and Netlify (possibly for security reasons) doesn't follow redirects on non-Netlify properties when doing the proxying.

In other words, I think this might go away if we just built this within Netlify.  (Note that _doesn't_ necessitate converting to Gatsby, necessarily!)

**Comment by glasser** [comment]
Filed https://github.com/apollo-specs/core/pull/6 to improve the core spec's handling of trailing slashes.

Would still be nice if support showed us an easy way to hide the backing URL better though :)

**Comment by abernix** [comment]
Netlify support was very helpful, and they looked through this and confirmed that the problem is as I suspected: GitHub is doing the redirect to `/` and Netlify doesn't continue to proxy through to redirects on external domains (for security reasons).

Why is GitHub doing the redirect from the unslashed to the slash?  As noted above, it [happens](https://apollo-specs.github.io/core/v0.1) without the slash on GitHub directly.  I'd venture a confident guess that it's because [we're publishing into a directory](https://github.com/apollo-specs/join/blob/main/.github/workflows/push-gh-pages.yml#L40-L42); the specification we're publishing isn't just a file, it's a directory.  The trailing slash makes sense, I think, and GitHub probably produces more reliably anchored relative links on pages that refer to it (e.g., `./page` on `site.com/v0.1` resolves to `site.com/page` whereas `./page` on `site/v0.1/` resolves to `site.com/v0.1/page`).  Anyhow, I can confirm this is fixed with #6, which I verified through https://specs.apollo.dev/jesse-test/core/v0.1 and https://specs.apollo.dev/jesse-test/core/v0.1/.
